### PR TITLE
8338981: Access to private classes should be permitted inside the permits clause of the enclosing top-level class

### PR DESCRIPTION
**Pull request by Evemose**
Fix involves adding new flag to after context that indicates that currently resolved symbol is in permits clause

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8341669](https://bugs.openjdk.org/browse/JDK-8341669) to be approved

### Issues
 * [JDK-8338981](https://bugs.openjdk.org/browse/JDK-8338981): Access to private classes should be permitted inside the permits clause of the enclosing top-level class (**Bug** - P3)
 * [JDK-8341669](https://bugs.openjdk.org/browse/JDK-8341669): Access to private classes and interfaces should be permitted inside the permits clause of the enclosing class or interface (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) Review applies to [441adf9f](https://git.openjdk.org/jdk/pull/20718/files/441adf9f3fbf6f57971e7ffae49b5786920847a2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20718/head:pull/20718` \
`$ git checkout pull/20718`

Update a local copy of the PR: \
`$ git checkout pull/20718` \
`$ git pull https://git.openjdk.org/jdk.git pull/20718/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20718`

View PR using the GUI difftool: \
`$ git pr show -t 20718`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20718.diff">https://git.openjdk.org/jdk/pull/20718.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20718#issuecomment-2311162173)
</details>

### CODE REVIEW

**Comment by lahodaj** [comment]
Others may have a different opinion, but this seems a bit too hacky to me (given we would need to maintain this for a long time).

I would suggest to try to add a new flag into `AttrContext`, along the lines of `permitPrivateAccessInHeader`, and set it while attributing the permitted types. Not sure how invasive that'd be, but hopefully not terribly much. And it should be clear (and documented) what's happening.

**Comment by Evemose** [comment]
> Others may have a different opinion, but this seems a bit too hacky to me (given we would need to maintain this for a long time).
> 
> I would suggest to try to add a new flag into `AttrContext`, along the lines of `permitPrivateAccessInHeader`, and set it while attributing the permitted types. Not sure how invasive that'd be, but hopefully not terribly much. And it should be clear (and documented) what's happening.

Honestly, I agree with you, I just didnt know how much invasive changes to existing code could be. I will look into AttrContext later today

**Comment by vicente-romero-oracle** [comment]
> > Others may have a different opinion, but this seems a bit too hacky to me (given we would need to maintain this for a long time).
> > I would suggest to try to add a new flag into `AttrContext`, along the lines of `permitPrivateAccessInHeader`, and set it while attributing the permitted types. Not sure how invasive that'd be, but hopefully not terribly much. And it should be clear (and documented) what's happening.
> 
> I have decided to write tests before reworking fix. Could you please suggest where could i put it? I have found tools/javac/diags/examples folder, but it seems to only contain negative tests, while I need positive ones. Is there any suitable place for this kind of thing?

most tests for sealed classes are under:
`test/langtools/tools/javac/sealed/`

**Comment by Evemose** [comment]
@lahodaj could you review please?

**Review by vicente-romero-oracle — COMMENTED** [review]
I think this change needs more test cases so that we can be sure that we are testing most of the possible combinations

**Comment by Evemose** [comment]
> I think this change needs more test cases so that we can be sure that we are testing most of the possible combinations

I tried to imagine some test cases where something could go wrong. Not really sure how to verify everything concidering the fact that permits is the last clause before attributing class code, which is allowed to access any private members, but I added some test cases veryfying everything from a different angles

**Comment by jddarcy** [comment]
If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.

**Comment by Evemose** [comment]
> If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.

Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 1]
```diff
@@ -458,6 +462,7 @@ public boolean isAccessible(Env<AttrContext> env, Type site, Symbol sym, boolean
             return isAccessible(env, site, checkInner) && notOverriddenIn(site, sym);
         }
     }
+
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:465]: Nit: please revert this change.

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            """
+            public sealed class T permits S.A {
+                private static final class A extends T {}
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"), root.resolve("T.java"))
+            .run(toolbox.Task.Expect.FAIL);
```
Comment (lahodaj) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:92]: Negative tests should, unless there is a very good reason not to, verify the exact errors printed. Use `-XDrawDiagnostics` option, and `.getOutputLines(Task.OutputKind.DIRECT)` on the result to get the output lines. (And, preferably also `.writeAll()` on all `toolbox.JavacTask` runs, passing or failing.) See for example here:

https://github.com/openjdk/jdk/blob/855c8a7def21025bc2fc47594f7285a55924c213/test/langtools/tools/javac/ImportModule.java#L175

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 3]
```diff
@@ -423,6 +423,10 @@ public boolean isAccessible(Env<AttrContext> env, Type site, Symbol sym, boolean
         case PRIVATE:
             return
                 (env.enclClass.sym == sym.owner // fast special case
+                 ||
+                 env.info.isPermitsClause
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:427]: I would suggest to have parentheses around the new test, i.e. around `env.info.isPermitsClause && ((JCClassDecl) env.tree).sym.outermostClass() == sym.owner.outermostClass()`, to make it clearer what is happening.

I would also suggest to reorder the tests, so that this new test is after the existing `env.enclClass.sym.outermostClass() == sym.owner.outermostClass()` (even at the cost of modifying the existing line to remove the `)`. It seems unnecessary to make almost all code go through the `env.info.isPermitsClause` check. I don't think there will be any real runtime difference, but still better to do the more common tests first. It is also, I think, better for reading the code to have the common cases earlier, unless there's another reason for the ordering.

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassOfOtherTopLevelClassInPermitsClause_whenCompiling_thenShouldFail() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
```
Comment (lahodaj) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:74]: Nit, up to you if you want to make the change: the `writeJavaFile` takes multiple `String`s, and can generate any number of files under the given directory. So the two calls to `writeJavaFile` can be coalesced to one.

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.nio.file.Path;
```
Comment (lahodaj) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:24]: Nit: we usually (try to) put the jtreg tags/header right after the license header (if there's any), and put imports after that. The import is part of the test, the jtreg header is test configuration.

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+/**
+ * @test
+ * @bug 8338981
+ * @summary Access to private classes should be permitted inside the permits clause of the enclosing top-level class
+ * @library /tools/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ * jdk.compiler/com.sun.tools.javac.main
+ * jdk.compiler/com.sun.tools.javac.util
```
Comment (lahodaj) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:33]: Nit:
```suggestion
 * @modules jdk.compiler/com.sun.tools.javac.api
 *                     jdk.compiler/com.sun.tools.javac.main
 *                     jdk.compiler/com.sun.tools.javac.util
```

**Review by lahodaj — COMMENTED** [review]
Overall, I think this is good. Left some comments to make the test and style similar to what is generally used in javac, at least in my opinion. Thanks for doing this change!

**Comment by lahodaj** [comment]
> > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> 
> Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?

In JBS, in the bug, there is "More", and if you click on that, there is "Create CSR". Click on that, and fill in the details in the text, as described in the template. When reviewing, I'll try to fill any flags that need to be filled (so that you know what gets filled next time).

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/AttrContext.java [hunk 7]
```diff
@@ -92,6 +92,10 @@ public class AttrContext {
      */
     boolean allowProtectedAccess = false;
 
+    /** Are we attributing a permits clause?
+     */
+    boolean isPermitsClause = false;
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/AttrContext.java:97]: you need to copy the value of this field in the ::dup method as it is done for other fields in this class

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java [hunk 8]
```diff
@@ -979,11 +979,13 @@ private void fillPermits(JCClassDecl tree, Env<AttrContext> baseEnv) {
             if (sym.isPermittedExplicit) {
                 ListBuffer<Symbol> permittedSubtypeSymbols = new ListBuffer<>();
                 List<JCExpression> permittedTrees = tree.permitting;
+                baseEnv.info.isPermitsClause = true;
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java:982]: usually we save the previous value of the field before setting a new value, then inside a `try` we invoke the code that should be aware of the field and then in the `finally` we restore the previous value of the field, see for example how we do for field `AttrContext::ctorPrologue`

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    public static void main(String... args) throws Exception {
+        new PrivateMembersInPermitClause().runTests();
+    }
+
+    public void runTests() throws Exception {
+        runTests(_ -> new Object[] {});
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClause_whenCompiling_thenShouldCompile() throws Exception {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:59]: I would use shorter method names, you can probably add more details about what the test does using javadoc or an internal comment

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassOfOtherTopLevelClassInPermitsClause_whenCompiling_thenShouldFail() throws Exception {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:75]: this test is failing for me

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClause_whenThanCompilingOtherClassThatReferencesPrivateClassInPermitsClause_thenShouldFail() throws Exception {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:146]: this one is failing for me too

Comment (Evemose) [reply test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:146]: It may be. I run heavy machine learning model on y pc right now, so i prefer to run tests using CI to not blow up my PC. I hope this doesnt bother anyone too much so I can figure out correct error messages that are expected in a few iterations.

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java (hunk 9 above)
Comment (Evemose) [reply test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:59]: it may be the way, but comments could become danglig over time, not sure if its the way. If I were writing with junit, I would use @DisplayName, not sure if there is something like this in jtreg. I will se what can be done though

**Comment by jddarcy** [comment]
> > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> 
> Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?

The CSR process is documented at https://wiki.openjdk.org/display/csr/Main. To create a CSR from a JBS issue, go to "More -> Create CSR."

Myself and the other reviewers can help guide you in the process.

**Comment by Evemose** [comment]
> > > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> > 
> > 
> > Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?
> 
> The CSR process is documented at https://wiki.openjdk.org/display/csr/Main. To create a CSR from a JBS issue, go to "More -> Create CSR."
> 
> Myself and the other reviewers can help guide you in the process.

I`m sorry, but I can not find any "More" button on bug page. Do I need an account in JBS in order to file CSR (i assume do since on page load there is a blinking second button new "Open" one that asks me to log in)? I am not an Oracle employee so I don't have one, and I am also not the one to report this issue

**Comment by jddarcy** [comment]
> > > > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> > > 
> > > 
> > > Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?
> > 
> > 
> > The CSR process is documented at https://wiki.openjdk.org/display/csr/Main. To create a CSR from a JBS issue, go to "More -> Create CSR."
> > Myself and the other reviewers can help guide you in the process.
> 
> I`m sorry, but I can not find any "More" button on bug page. Do I need an account in JBS in order to file CSR (i assume do since on page load there is a blinking second button new "Open" one that asks me to log in)? I am not an Oracle employee so I don't have one, and I am also not the one to report this issue

Yes; someone with a JBS account will need to create the CSR.

**Comment by vicente-romero-oracle** [comment]
> > > > > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> > > > 
> > > > 
> > > > Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?
> > > 
> > > 
> > > The CSR process is documented at https://wiki.openjdk.org/display/csr/Main. To create a CSR from a JBS issue, go to "More -> Create CSR."
> > > Myself and the other reviewers can help guide you in the process.
> > 
> > 
> > I`m sorry, but I can not find any "More" button on bug page. Do I need an account in JBS in order to file CSR (i assume do since on page load there is a blinking second button new "Open" one that asks me to log in)? I am not an Oracle employee so I don't have one, and I am also not the one to report this issue
> 
> Yes; someone with a JBS account will need to create the CSR.

I have just created a blank one: https://bugs.openjdk.org/browse/JDK-8341669

**Comment by Evemose** [comment]
> > > > > > If this fix is proposing to align javac behavior with the JLS, while no spec change would be needed. A CSR would need to be filed for the behavioral change.
> > > > > 
> > > > > 
> > > > > Hello! I am not familiar with CSR process. I tried to find a way to file it using java bug database, but could not find anything. Could you tell me what should I do and where, please?
> > > > 
> > > > 
> > > > The CSR process is documented at https://wiki.openjdk.org/display/csr/Main. To create a CSR from a JBS issue, go to "More -> Create CSR."
> > > > Myself and the other reviewers can help guide you in the process.
> > > 
> > > 
> > > I`m sorry, but I can not find any "More" button on bug page. Do I need an account in JBS in order to file CSR (i assume do since on page load there is a blinking second button new "Open" one that asks me to log in)? I am not an Oracle employee so I don't have one, and I am also not the one to report this issue
> > 
> > 
> > Yes; someone with a JBS account will need to create the CSR.
> 
> I have just created a blank one: https://bugs.openjdk.org/browse/JDK-8341669, I will write the text of the CSR

Thanks. I have changed bugid in test and pr

**Comment by mcimadamore** [comment]
The fix looks good, and it seems to reflect the spirit of the JLS.

**Review by mcimadamore — APPROVED** [review]

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm, thanks for fixing this

**Comment by vicente-romero-oracle** [comment]
I need a reviewer for the CSR, thanks!

**Comment by lahodaj** [comment]
I am sorry for not fully realizing this sooner, I forgot about it. But, the JLS older than 19 didn't have this provision for the nested private classes, I believe. Do we need to limit this to only `-source 19+`?

**Comment by vicente-romero-oracle** [comment]
> I am sorry for not fully realizing this sooner, I forgot about it. But, the JLS older than 19 didn't have this provision for the nested private classes, I believe. Do we need to limit this to only `-source 19+`?

good catch, yes I agree

**Comment by Evemose** [comment]
> > I am sorry for not fully realizing this sooner, I forgot about it. But, the JLS older than 19 didn't have this provision for the nested private classes, I believe. Do we need to limit this to only `-source 19+`?
> 
> good catch, yes I agree. @Evemose I think that we will need to add a new feature, to `com.sun.tools.javac.code.Source.Feature`. This PR is a recent-ish example: `https://github.com/openjdk/jdk/pull/12893`

I have added feature, but would apreciate any advice on how to improve error reporting. You could see currecnt behaviour in added test, but I feel like it could be improved

**Comment by Evemose** [comment]
> > > I am sorry for not fully realizing this sooner, I forgot about it. But, the JLS older than 19 didn't have this provision for the nested private classes, I believe. Do we need to limit this to only `-source 19+`?
> > 
> > 
> > good catch, yes I agree. @Evemose I think that we will need to add a new feature, to `com.sun.tools.javac.code.Source.Feature`. This PR is a recent-ish example: `https://github.com/openjdk/jdk/pull/12893`
> 
> I have added feature, but would apreciate any advice on how to improve error reporting. You could see currecnt behaviour in added test, but I feel like it could be improved, specifically position reporting and erroneous symbol reporting, but I am not really sure how its dont properly

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 12]
```diff
@@ -458,6 +462,19 @@ public boolean isAccessible(Env<AttrContext> env, Type site, Symbol sym, boolean
             return isAccessible(env, site, checkInner) && notOverriddenIn(site, sym);
         }
     }
+
+    private boolean privateMemberInPermitsClauseIfAllowed(Env<AttrContext> env, Symbol sym) {
+        if (env.info.isPermitsClause && ((JCClassDecl) env.tree).sym.outermostClass() == sym.owner.outermostClass()) {
+            if (allowPrivateMembersInPermitsClause) {
+                return true;
+            } else {
+                log.error(env.tree.pos(), Errors.PrivateMembersInPermitsClause);
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:471]: we don't issue a compiler error while checking if a symbol is accessible or not, it should be the caller's responsibility to decide if a compiler error should be issued or not as it is done in other cases

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 13]
```diff
@@ -458,6 +462,19 @@ public boolean isAccessible(Env<AttrContext> env, Type site, Symbol sym, boolean
...
+                return false;
+            }
+        }
+        return false;
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:475]: shouldn't we return true here?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java (hunk 12 above)
Comment (Evemose) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: thanks for the note, will keep in mind from now on

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java (hunk 13 above)
Comment (Evemose) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:470]: No, method name is privateMemberInPermitsClause if allowed, and return is:
1. If condition in if is met, then return whether feature is enabled or not (value of field allowPrivateMembersInPermitsClause)
2. If condition is not met, then this is NOT privateMemberInPermitsClause, so return false

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:470]: what about:
```
private boolean privateMemberInPermitsClauseIfAllowed(Env<AttrContext> env, Symbol sym) {
    return allowPrivateMembersInPermitsClause &&
           env.info.isPermitsClause &&
           ((JCClassDecl) env.tree).sym.outermostClass() == sym.owner.outermostClass();
}
```

File: src/jdk.compiler/share/classes/com/sun/tools/javac/resources/compiler.properties [hunk 14]
```diff
@@ -3247,6 +3247,9 @@ compiler.misc.feature.flexible.constructors=\
 compiler.misc.feature.module.imports=\
     module imports
 
+compiler.misc.feature.private.members.in.permits.clause=\
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/resources/compiler.properties:3250]: this one is not used now, can be removed

**Comment by Evemose** [comment]
By the way, does anyone know why ci fails? It fails on generate report, not on tests. I tried to sync branch with master but it seems like it has not had any affect

File: test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    /**
+     * Tests that a private class from another top-level class in the permits clause fails to compile.
+     */
+    @Test
+    public void otherTopLevelPrivateClassFails() throws Exception {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:80]: this test is failing for me:

```
Expected errors: [S.java:5:25: compiler.err.report.access: S.A, private, S, 1 error], 
but got: [S.java:4:25: compiler.err.report.access: S.A, private, S, 1 error]
```

Comment (Evemose) [reply test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java:80]: Thanks, I forgot to change the line number when removed empty line for jcheck

**Comment by vicente-romero-oracle** [comment]
@Evemose I was waiting for the results of the tests after your last change, all green, we are good to go

**Review by vicente-romero-oracle — APPROVED** [review]
thanks!

**Comment by Evemose** [comment]
> @Evemose I was waiting for the results of the tests after your last change, all green, we are good to go

Glad we are finally done with this one. Got a bit busy lately, pr could`ve taken much less time, thanks for your patience. WOuld you mind sponsoring?

**Comment by vicente-romero-oracle** [comment]
> > @Evemose I was waiting for the results of the tests after your last change, all green, we are good to go
> 
> Glad we are finally done with this one. Got a bit busy lately, pr could`ve taken much less time, thanks for your patience. Would you mind sponsoring?

thanks to you!
